sklearn/linear_model/glm.py

```diff
@@ -0,0 +1,872 @@
+    def __init__(self):
+        super(PoissonDistribution, self).__init__(power=1)
+
+class GammaDistribution(TweedieDistribution):
+    """Class for the Gamma distribution"""
+    def __init__(self):
+        super(GammaDistribution, self).__init__(power=2)
+
+class InverseGaussianDistribution(TweedieDistribution):
+    """Class for the scaled InverseGaussianDistribution distribution"""
+    def __init__(self):
+        super(InverseGaussianDistribution, self).__init__(power=3)
+
+class GeneralizedHyperbolicSecand(ExponentialDispersionModel):
+    """A class for the von Generalized Hyperbolic Secand (GHS) distribution.
+
+    The GHS distribution is for data y in (-inf, inf).
+    """
+    def __init__(self):
+        self._lower_bound = -np.Inf
+        self._upper_bound = np.Inf
+
+    @property
+    def lower_bound(self):
+        return self._lower_bound
+
+    @property
+    def upper_bound(self):
+        return self._upper_bound
+
+    def in_y_range(self, x):
+        np.logical_and(
+            np.greater(x, self.lower_bound),
+            np.less(x, self.lower_bound)
+            )
+
+    def unit_variance(self, mu):
+        return 1 + mu**2
+
+    def unit_variance_derivative(self, mu):
+        return 2*mu
+
+    def unit_deviance(self, y, mu):
+        return (2*y*(np.arctan(y) - np.arctan(mu))
+            + np.log((1+mu**2)/(1+y**2)))
+
+
+
+class GeneralizedLinearModel(LinearModel, RegressorMixin):
+    """
+    Class to fit a Generalized Linear Model (GLM) based on reproductive
+    Exponential Dispersion Models (EDM).
+
+    Assumptions:
+
+    - The target values y_i are realizations of random variables
+      :math:`Y_i \sim \mathrm{EDM}(\mu_i, \frac{\phi}{w_i})` with dispersion
+      parameter :math:`\phi` and weights :math:`w_i`.
+    - The expectation of :math:`Y_i` is :math:`mu_i=\mathrm{E}[Y]=h(\eta_i)`
+      whith the linear predictor :math:`\eta=X*\beta`, inverse link function
+      :math:`h(\eta)`, design matrix :math:`X` and parameters :math:`\beta`
+      to be estimated.
+
+    Note that the first assumption implies
+    :math:`\mathrm{Var}[Y_i]=\frac{\phi}{w_i} v(\mu_i)` with uni variance
+    function :math:`v(\mu)`.
+
+    The fit itself does not need Y to be from an EDM, but only assumes
+    the first two moments :math:`E[Y_i]=\mu_i=h(\eta_i)` and
+    :math:`Var[Y_i]=\frac{\phi}{w_i} v(\mu_i)`
+
+    The parameters :math:`\beta` are estimated by maximum likelihood which is
+    equivalent to minimizing the deviance.
+
+    TODO: Estimation of the dispersion parameter phi.
+
+    TODO: Notes on 'scaled' Poisson and weights
+
+    Parameters
+    ----------
+    fit_intercept : boolean, optional, default True
+        whether to calculate the intercept for this model. If set
+        to False, no intercept will be used in calculations
+        (e.g. data is expected to be already centered).
+
+    family : ExponentialDispersionModel, optional, default NormalDistribution()
```

Comment:
Could you please have this take a string, or a model only when necessary? Perhaps you can add an additional parameter or for power in the general Tweedie case.

In turn, the models with string representations should be private. I don't think we want to litter linear_model with models and links.